test/modules/EdgeHubRestartTester/src/Program.cs

```diff
@@ -0,0 +1,166 @@
+                        Settings.Current.TransportType,
+                        ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
+                        ModuleUtil.DefaultTransientRetryStrategy,
+                        Logger);
+                }
+
+                DateTime testStart = DateTime.UtcNow;
+                DateTime testExpirationTime = testStart + Settings.Current.TestDuration;
+
+                while ((!cts.IsCancellationRequested) && (DateTime.UtcNow < testExpirationTime))
+                {
+                    (DateTime restartTime, _) = await RestartModules(iotHubServiceClient);
+                    DateTime eachTestExpirationTime = restartTime.Add(Settings.Current.RestartPeriod);
+
+                    // Increment the counter when issue an edgeHub restart
+                    restartCount++;
+
+                    // Setup Message Task
+                    Task sendMessageTask = Task.CompletedTask;
+                    if (Settings.Current.MessageEnabled)
+                    {
+                        sendMessageTask =
+                            new MessageEdgeHubConnector(
+                                msgModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                restartTime,
+                                Logger,
+                                cts.Token)
+                            .StartAsync();
+                    }
+
+                    // Setup Direct Method Task
+                    Task directMethodTask = Task.CompletedTask;
+                    if (Settings.Current.DirectMethodEnabled)
+                    {
+                        directMethodTask =
+                            new DirectMethodEdgeHubConnector(
+                                dmModuleClient,
+                                batchId,
+                                eachTestExpirationTime,
+                                restartTime,
+                                restartCount,
+                                Logger,
+                                cts.Token)
+                            .StartAsync();
+                    }
+
+                    // Wait for the two task to be done before do a restart
+                    await Task.WhenAll(new[] { sendMessageTask, directMethodTask });
+
+                    // Wait until the specified restart period to do another restart
+                    await Task.Delay((int)(eachTestExpirationTime - DateTime.UtcNow).TotalMilliseconds, cts.Token);
+                }
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Exception caught: {e}");
+                throw;
+            }
+            finally
+            {
+                iotHubServiceClient?.Dispose();
+                dmModuleClient?.Dispose();
+                msgModuleClient?.Dispose();
+            }
+
+            await cts.Token.WhenCanceled();
```

Comment:
Will this be automatically cancelled when the test run ends?

Reply:
Yes